src/main/host/syscall/unistd.rs

```diff
@@ -84,19 +106,34 @@ pub fn read(sys: &mut c::SysCallHandler, args: &c::SysCallArgs) -> c::SysCallRet
     let fd = unsafe { args.args[0].as_i64 } as libc::c_int;
     let buf_ptr = unsafe { args.args[1].as_ptr };
     let buf_size = unsafe { args.args[2].as_u64 } as libc::size_t;
-    let offset = 0 as libc::off_t;
+    let offset = 0;
+
+    // get the descriptor, or return early if it doesn't exist
+    let desc = match syscall::get_descriptor(fd, sys.process) {
+        Ok(d) => unsafe { &mut *d },
+        Err(errno) => return SyscallReturn::Error(errno).into(),
+    };
+
+    // if it's a legacy descriptor, use the C syscall handler instead
+    let desc = match desc {
+        CompatDescriptor::New(d) => d,
+        CompatDescriptor::Legacy(_) => unsafe {
+            return c::syscallhandler_read(
+                sys as *mut c::SysCallHandler,
+                args as *const c::SysCallArgs,
+            );
+        },
+    };
 
-    read_helper(sys, args, fd, buf_ptr, buf_size, offset)
+    read_helper(sys, fd, desc, buf_ptr, buf_size, offset)
```

Comment:
control flow might be a bit easier to follow as:

```
return match desc {
  CompatDescriptor::New(d) => read_helper(...),
  CompatDescriptor::Legacy(_) => unsafe { c::syscallhandler_read(...) }
}
```

Reply:
Good idea, I made this change.